### ISSUE TITLE
Unhandled promise rejection

### ISSUE THREAD

**Issue by rogierschouten** [poelstra/ts-stream#31]
The following code shows an unhandled rejection and it shouldn't. Note that extra's like 'delay' are present to test for proper waiting on result promises.

```typescript

var promises = require("@exlent/promises");

process.on("unhandledRejection", (reason, p) => {
	console.log("unhandled:", reason)
});

process.on("rejectionHandled", (reason, p) => {
	console.log("re-handled:", reason)
});

var s = require("ts-stream");

const result = new s.Stream();
const stream = new s.Stream();
stream.end(new Error("foo"))
	.catch((error) => undefined);

function delay(v) {
	return promises.delay(1000).then(() => { if (v) throw v } );
}

stream.forEach(
	(v) => result.write(v),
	(err) => result.end(err, stream.result().then(delay, delay))
);


result.forEach(() => undefined, (e) => {
	console.log("ended 1");
	throw e;
})
	.then(() => console.log("then"), () => console.log("caught"));

console.log("start");
setTimeout(() => undefined, 5000);
```

**Comment by poelstra** [comment]
An additional bug found during this investigation (and fixed in #32), was that a downstream stream doesn't properly wait for upstream to completely end (i.e. its `result()` to resolve or reject) in case the downstream ender threw an error or returned a rejection.

An explicit test for this is added as well now.

**Comment by poelstra** [comment]
Published as 1.0.1.
